$OpenBSD: patch-src_main_c,v 1.7 2012/04/11 08:41:58 dcoppa Exp $

Warn if ctl is not a fifo.
Ordinary files are not supported and cause problems, so close the
fd and print a warning instead.
(upstream git commit ed6d013054f589f2999926b02f4ac6dc76c7fe91)

--- src/main.c.orig	Wed Apr 11 10:33:34 2012
+++ src/main.c	Wed Apr 11 10:32:20 2012
@@ -369,9 +369,19 @@ int main (int argc, char **argv) {
 	assert (sizeof (app.input.fds) / sizeof (*app.input.fds) >= 2);
 	app.input.fds[1] = open (app.settings.fifo, O_RDWR);
 	if (app.input.fds[1] != -1) {
-		FD_SET(app.input.fds[1], &app.input.set);
-		BarUiMsg (&app.settings, MSG_INFO, "Control fifo at %s opened\n",
-				app.settings.fifo);
+		struct stat s;
+
+		/* check for file type, must be fifo */
+		fstat (app.input.fds[1], &s);
+		if (!S_ISFIFO (s.st_mode)) {
+			BarUiMsg (&app.settings, MSG_ERR, "File at %s is not a fifo\n", app.settings.fifo);
+			close (app.input.fds[1]);
+			app.input.fds[1] = -1;
+		} else {
+			FD_SET(app.input.fds[1], &app.input.set);
+			BarUiMsg (&app.settings, MSG_INFO, "Control fifo at %s opened\n",
+					app.settings.fifo);
+		}
 	}
 	app.input.maxfd = app.input.fds[0] > app.input.fds[1] ? app.input.fds[0] :
 			app.input.fds[1];
